### PR TITLE
pauli refactor

### PR DESCRIPTION
**Pull request by Bchass**
## Description
Closes: https://github.com/vprusso/toqito/issues/529

I left `is_sparse` in for the moment even though it's not technically being used in `pauli.py` due to the test cases relying on True or False. Would it make sense to completely remove `is_sparse` and drop `True` from the test cases? @purva-thakre  I believe so, but drafting this for PR for now.

## Changes
  -  [x] Remove `csr_matrix`

## Checklist

  -  [x] Use `ruff` and `pylint` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/matrices/pauli.py
Comment (purva-thakre) [inline toqito/matrices/pauli.py:96]: @Bchass I think instead of removing `is_sparse` option  completely due to `sparse.csr_matrix`, we could try to switch to using `sparse.csr_array` instead. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_array.html#scipy.sparse.csr_array

I haven't tried this out yet, FYI! If you run into a lot of issues, then we could simply remove that option in case we can't find other options.

Comment (Bchass) [reply toqito/matrices/pauli.py:96]: Looked into `sparse.csr_array` and it works a lot better for this use case. No need to alter any test cases.

File: toqito/matrices/pauli.py
Comment (purva-thakre) [inline toqito/matrices/pauli.py:9]: Why remove `np.ndarray`? You get a `csr_array` for the output only when we provide `is_sparse` is not `False`.

Comment (Bchass) [reply toqito/matrices/pauli.py:9]: I was thinking back to: https://github.com/vprusso/toqito/pull/521#discussion_r1543072900 and that's why initially I removed `is_sparse` and made some changes to the test cases. But I'll wait for a further response.

Comment (purva-thakre) [reply toqito/matrices/pauli.py:9]: Yes, but `issparse` is a method for both sparse array and sparse matrix types. So, the structure of the function shouldn't change that much. Before, the return type was an array or a sparse matrix. Now, it can be an `np.ndarray` for all options of allowed inputs. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.issparse.html#scipy.sparse.issparse

**Review by purva-thakre — APPROVED** [review]
LGTM! I'll wait for @vprusso to Ok this before merging.

**Pull request merged**
